src/CSharpCollections/DoubleEndedQueue/CircularDoubleEndedQueue.cs

```diff
@@ -0,0 +1,513 @@
+﻿using System.Collections.Generic;
+
+namespace System.Collections.SharpCollect
+{
+    /// <summary>
+    /// A Double Ended Queue allows for Enqueue-ing and Dequeue-ing from the front and back
+    /// but always keeps the same capacity
+    /// </summary>
+    /// <typeparam name="T">The type of object stored in the queue</typeparam>
+    public class CircularDoubleEndedQueue<T> : IDoubleEndedQueue<T>, ICollection
+    {
+        // The items begin stored
+        private T[] _items;
+        // The position of the next item to be popped
+        private int _indexFront = 0;
+        // The position of the next item to be pushed
+        private int _indexBack = 0;
+
+        // Locking for multi-threaded operations
+        private object _syncRoot;
+
+        /// <summary>
+        /// Initializes a Double Ended Queue with a capacity
+        /// </summary>
+        /// <param name="capacity">The number of items before the queue attempts to grow</param>
+        /// <exception cref="ArgumentException">Thrown if the specified capacity is invalid</exception>
+        public CircularDoubleEndedQueue(int capacity) 
+        {
+            if (capacity < 3) { throw new ArgumentOutOfRangeException(nameof(capacity),"Initial capacity must be greater than 2", nameof(capacity)); }
+            _items = new T[capacity];
+        }
+
+        /// <summary>
+        /// Parameter-less constructor
+        /// </summary>
+        /// <remarks>
+        /// The default size is 32
+        /// </remarks>
+        public CircularDoubleEndedQueue() : this(32) { }
+
+        /// <summary>
+        /// The number of items in the queue
+        /// </summary>
+        public int Capacity => _items.Length;
+
+        /// <summary>
+        /// The number of items in the queue
+        /// </summary>
+        public int Count { get; private set; } = 0;
+
+        /// <summary>
+        /// Is this Collection synchronized (thread-safe)?
+        /// </summary>
+        public bool IsSynchronized => false;
+
+        /// <summary>
+        /// Synchronization root for this object.
+        /// </summary>
+        public object SyncRoot
+        {
+            get
+            {
+                if (_syncRoot == null)
+                {
+                    System.Threading.Interlocked.CompareExchange<Object>(ref _syncRoot, new Object(), null);
+                }
+                return _syncRoot;
+            }
+        }
+
+        /// <summary>
+        /// Returns true if the list can be written to
+        /// </summary>
+        public bool IsReadOnly => false;
+
+
+        /// <summary>
+        /// Adds obj to the back/tail/right of the queue
+        /// </summary>
+        /// <param name="obj">The object to add</param>
+        public void Enqueue(T obj)
+        {
+            _items[_indexBack++] = obj;
```

Comment:
Are you wanting the exception to be thrown if index exceeds max capacity?

Reply:
Comment added